[AIRFLOW-4456] Add sub-classable BaseBranchOperator

chronitis opened the pull request:
### Jira

- [x] My PR addresses https://issues.apache.org/jira/browse/AIRFLOW-4456

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

This adds an abstract BranchOperator, which implementors can use to
create concrete operators for branching workflows fully encapsulated in
a class, rather than needing to pass an external callable (or
self-reference) to BranchPythonOperator. Also adds tests for same.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

`tests/operators/test_branch_operator.py`

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

Commit:
[AIRFLOW-4456] Add BranchOperator

This adds an abstract BranchOperator, which implementors can use to
create concrete operators for branching workflows fully encapsulated in
a class, rather than needing to pass an external callable (or
self-reference) to BranchPythonOperator. Also adds tests for same.

Commit:
[AIRFLOW-4456] Refactor branching logic into SkipMixin

This moves the logic of skipping non-selected branches into
airflow.models.SkipMixin:skip_all_except, and modifies
BranchPythonOperator and BranchOperator to use this mixin instead of
implementing the logic themselves.

Commit:
[AIRFLOW-4456] Document BranchOperator

Adds a brief section on BranchOperator to docs/concepts.rst

ashb commented on airflow/operators/branch_operator.py:
```diff
@@ -0,0 +1,51 @@
+from typing import Union, Iterable, Dict
+
+
+class BranchOperator(BaseOperator, SkipMixin):
```
```suggestion
class BaseBranchOperator(BaseOperator, SkipMixin):
```

Just to make it a tiny bit clearer that this is intended to be sub-classed (i.e. not used in a DAG, similar to BaseOperator)

ashb reviewed (changes requested):
Small naming change, but yes this looks good, thanks!

ashb commented on docs/concepts.rst:
```diff
@@ -581,6 +581,26 @@ For example:
   start_op >> branch_op >> [continue_op, stop_op]
 
+If you wish to implement your own operators with branching functionality,
+you can inherit from ``BranchOperator``, which behaves similarly to
```
```suggestion
you can inherit from :class:`~airflow.operators.branch_operator.BaseBranchOperator`, which behaves similarly to
```

ashb commented on docs/concepts.rst:
```diff
@@ -581,6 +581,26 @@ For example:
 
+If you wish to implement your own operators with branching functionality,
+you can inherit from ``BranchOperator``, which behaves similarly to
+``BranchPythonOperator`` but expects you to provide an implementation of the
```
```suggestion
:class:`~airflow.operators.python_operator.BranchPythonOperator` but expects you to provide an implementation of the
```

ashb commented on docs/concepts.rst:
```diff
@@ -581,6 +581,26 @@ For example:
+If you wish to implement your own operators with branching functionality,
+you can inherit from ``BranchOperator``, which behaves similarly to
+``BranchPythonOperator`` but expects you to provide an implementation of the
+method ``choose_branch``. As with the callable for ``BranchPythonOperator``,
```
```suggestion
method ``choose_branch``. As with the callable for BranchPythonOperator,
```

ashb reviewed (changes requested):
Improve doc linking to classes too.

The suggestions as they are will make the line-wrapping messy, so probably tidy that up too :)

Commit:
[AIRFLOW-4456] BranchOperator->BaseBranchOperator

BranchOperator is prefixed Base to reflect that it is designed to be
subclassed, and the documentation and tests updated accordingly.

ashb committed:
Update branch_operator.py
